### ISSUE TITLE
QueryFilter being lost after update to EF Core 5.0.2

### ISSUE THREAD

**Issue by fabiano-ferreira** [dotnet/efcore#23979]
Hello,

I have a problem happening. I'm using a Blazor Server Project with Repository pattern. My entities are configured to work with a soft delete and a query filter:

`modelBuilder.Entity<Employee>().HasQueryFilter(p => !p.Deleted);`

It generally works fine and the query filter is being applied without an issue.

However if my browser gets idle for some hours (I stopped using yesterday and used again today), when I apply the same query as before (same LINQ to Entities code), the query filter is not being applied. 

This started to happen after I updated to version 5.0.2.

My DBContext is configured to use a Scoped Lifetime.

Could you, please, help me?

Thanks!

Fabiano

**Comment by ajcvickers** [comment]
@fabiano-ferreira There isn't enough here for us to reproduce what you are seeing. If possible, please attach a small, runnable project or post a small, runnable code listing that reproduces what you are seeing so that we can investigate.

**Comment by fabiano-ferreira** [comment]
Ok...

Here is the code that is generating the resultset. Currently the repository has 86 records. From those, 44 are not deleted (Deleted field equals false).

` var userOrganizations = await _applicationUserRepository.GetAll()
                                        .Include(o => o.ApplicationUserOrganizations)
                                            .ThenInclude(o => o.Organization)
                                                .ThenInclude(o => o.OrganizationLocations).ThenInclude(o => o.Location)
                                        .Include(o => o.ApplicationUserOrganizations)
                                            .ThenInclude(o => o.Organization)
                                                .ThenInclude(o => o.Employees.OrderBy(l => l.LastName).ThenBy(f => f.FirstName))
                                        .Where(u => u.Id == userId)
                                    .SelectMany(o => o.ApplicationUserOrganizations)
                                    .Select(o => o.Organization)
                                    .ToListAsync();`
 
As you see I'm listing all the Organizations that have some relationship with the user. One of the included collections is the list of Employees in each organization.

This query runs perfectly if I just load the page for the first time. This is the query being generated (I got it using SQL Server Profiler):

`SELECT [t0].[Id], [t0].[Abbreviation], [t0].[Active], [t0].[CreatedBy], [t0].[CreatedDateTime], [t0].[Deleted], [t0].[MaxTests], [t0].[MinTests], [t0].[Name], [t0].[OrgGraphic], [t0].[ProviderId], [t0].[UpdatedBy], [t0].[UpdatedDateTime], [a].[Id], [t].[ApplicationUserId], [t].[OrganizationId], [t1].[OrganizationId], [t1].[LocationId], [t1].[CreatedBy], [t1].[CreatedDateTime], [t1].[Default], [t1].[Deleted], [t1].[UpdatedBy], [t1].[UpdatedDateTime], [t1].[Id], [t1].[AddressLine1], [t1].[AddressLine2], [t1].[City], [t1].[CreatedBy0], [t1].[CreatedDateTime0], [t1].[Deleted0], [t1].[Fax], [t1].[Name], [t1].[Phone], [t1].[PostalCode], [t1].[PrivateCollection], [t1].[Province], [t1].[UpdatedBy0], [t1].[UpdatedDateTime0], [t2].[Id], [t2].[CreatedBy], [t2].[CreatedDateTime], [t2].[DateOfBirth], [t2].[Deleted], [t2].[FirstName], [t2].[Gender], [t2].[LastName], [t2].[OrgEmployeeId], [t2].[OrganizationId], [t2].[Suspended], [t2].[UpdatedBy], [t2].[UpdatedDateTime]
FROM [AspNetUsers] AS [a]
INNER JOIN (
    SELECT [a0].[ApplicationUserId], [a0].[OrganizationId]
    FROM [ApplicationUserOrganization] AS [a0]
    WHERE [a0].[Deleted] <> CAST(1 AS bit)
) AS [t] ON [a].[Id] = [t].[ApplicationUserId]
INNER JOIN (
    SELECT [o].[Id], [o].[Abbreviation], [o].[Active], [o].[CreatedBy], [o].[CreatedDateTime], [o].[Deleted], [o].[MaxTests], [o].[MinTests], [o].[Name], [o].[OrgGraphic], [o].[ProviderId], [o].[UpdatedBy], [o].[UpdatedDateTime]
    FROM [Organization] AS [o]
    WHERE [o].[Deleted] <> CAST(1 AS bit)
) AS [t0] ON [t].[OrganizationId] = [t0].[Id]
LEFT JOIN (
    SELECT [o0].[OrganizationId], [o0].[LocationId], [o0].[CreatedBy], [o0].[CreatedDateTime], [o0].[Default], [o0].[Deleted], [o0].[UpdatedBy], [o0].[UpdatedDateTime], [l].[Id], [l].[AddressLine1], [l].[AddressLine2], [l].[City], [l].[CreatedBy] AS [CreatedBy0], [l].[CreatedDateTime] AS [CreatedDateTime0], [l].[Deleted] AS [Deleted0], [l].[Fax], [l].[Name], [l].[Phone], [l].[PostalCode], [l].[PrivateCollection], [l].[Province], [l].[UpdatedBy] AS [UpdatedBy0], [l].[UpdatedDateTime] AS [UpdatedDateTime0]
    FROM [OrganizationLocation] AS [o0]
    INNER JOIN [Location] AS [l] ON [o0].[LocationId] = [l].[Id]
    WHERE [o0].[Deleted] <> CAST(1 AS bit)
) AS [t1] ON [t0].[Id] = [t1].[OrganizationId]
LEFT JOIN (
    SELECT [e].[Id], [e].[CreatedBy], [e].[CreatedDateTime], [e].[DateOfBirth], [e].[Deleted], [e].[FirstName], [e].[Gender], [e].[LastName], [e].[OrgEmployeeId], [e].[OrganizationId], [e].[Suspended], [e].[UpdatedBy], [e].[UpdatedDateTime]
    FROM [Employee] AS [e]
    WHERE [e].[Deleted] <> CAST(1 AS bit)
) AS [t2] ON [t0].[Id] = [t2].[OrganizationId]
WHERE [a].[Id] = @__userId_0
ORDER BY [a].[Id], [t].[ApplicationUserId], [t].[OrganizationId], [t0].[Id], [t1].[OrganizationId], [t1].[LocationId], [t1].[Id], [t2].[LastName], [t2].[FirstName], [t2].[Id]`


However, if the Blazor server page gets idle for some hours, it seems that EF Core stops considering the query filter being mentioned before.

So, sometimes the code runs OK! And sometimes the resultset is wrong (it simply doesn't apply the query filter to ignore deleted records). This was not happening before version 5.0.2. It started happening recently.

**Comment by ajcvickers** [comment]
@fabiano-ferreira 

> However, if the Blazor server page gets idle for some hours, it seems that EF Core stops considering the query filter being mentioned before.

I don't think there we are going to be able to figure out why this is happening without more information. Have you set logging to Debug level and captured the logs over the time period when this happens? If so, can you share them?

**Comment by fabiano-ferreira** [comment]
Ok:

This is the query generated after at least 10 hours without a website interaction:

`SELECT [o].[Id], [o].[Abbreviation], [o].[Active], [o].[CreatedBy], [o].[CreatedDateTime], [o].[Deleted], [o].[MaxTests], [o].[MinTests], [o].[Name], [o].[OrgGraphic], [o].[ProviderId], [o].[UpdatedBy], [o].[UpdatedDateTime], [a].[Id], [a0].[ApplicationUserId], [a0].[OrganizationId], [t].[OrganizationId], [t].[LocationId], [t].[CreatedBy], [t].[CreatedDateTime], [t].[Default], [t].[Deleted], [t].[UpdatedBy], [t].[UpdatedDateTime], [t].[Id], [t].[AddressLine1], [t].[AddressLine2], [t].[City], [t].[CreatedBy0], [t].[CreatedDateTime0], [t].[Deleted0], [t].[Fax], [t].[Name], [t].[Phone], [t].[PostalCode], [t].[PrivateCollection], [t].[Province], [t].[UpdatedBy0], [t].[UpdatedDateTime0], [t0].[Id], [t0].[CreatedBy], [t0].[CreatedDateTime], [t0].[DateOfBirth], [t0].[Deleted], [t0].[FirstName], [t0].[Gender], [t0].[LastName], [t0].[OrgEmployeeId], [t0].[OrganizationId], [t0].[Suspended], [t0].[UpdatedBy], [t0].[UpdatedDateTime]
FROM [AspNetUsers] AS [a]
INNER JOIN [ApplicationUserOrganization] AS [a0] ON [a].[Id] = [a0].[ApplicationUserId]
INNER JOIN [Organization] AS [o] ON [a0].[OrganizationId] = [o].[Id]
LEFT JOIN (
    SELECT [o0].[OrganizationId], [o0].[LocationId], [o0].[CreatedBy], [o0].[CreatedDateTime], [o0].[Default], [o0].[Deleted], [o0].[UpdatedBy], [o0].[UpdatedDateTime], [l].[Id], [l].[AddressLine1], [l].[AddressLine2], [l].[City], [l].[CreatedBy] AS [CreatedBy0], [l].[CreatedDateTime] AS [CreatedDateTime0], [l].[Deleted] AS [Deleted0], [l].[Fax], [l].[Name], [l].[Phone], [l].[PostalCode], [l].[PrivateCollection], [l].[Province], [l].[UpdatedBy] AS [UpdatedBy0], [l].[UpdatedDateTime] AS [UpdatedDateTime0]
    FROM [OrganizationLocation] AS [o0]
    INNER JOIN [Location] AS [l] ON [o0].[LocationId] = [l].[Id]
) AS [t] ON [o].[Id] = [t].[OrganizationId]
LEFT JOIN (
    SELECT [e].[Id], [e].[CreatedBy], [e].[CreatedDateTime], [e].[DateOfBirth], [e].[Deleted], [e].[FirstName], [e].[Gender], [e].[LastName], [e].[OrgEmployeeId], [e].[OrganizationId], [e].[Suspended], [e].[UpdatedBy], [e].[UpdatedDateTime]
    FROM [Employee] AS [e]    
) AS [t0] ON [o].[Id] = [t0].[OrganizationId]
WHERE [a].[Id] = @__userId_0
ORDER BY [a].[Id], [a0].[ApplicationUserId], [a0].[OrganizationId], [o].[Id], [t].[OrganizationId], [t].[LocationId], [t].[Id], [t0].[LastName], [t0].[FirstName], [t0].[Id]`


And THIS is the query generated after I completely refresh the browser (hit enter on the browser address bar):

`SELECT [t0].[Id], [t0].[Abbreviation], [t0].[Active], [t0].[CreatedBy], [t0].[CreatedDateTime], [t0].[Deleted], [t0].[MaxTests], [t0].[MinTests], [t0].[Name], [t0].[OrgGraphic], [t0].[ProviderId], [t0].[UpdatedBy], [t0].[UpdatedDateTime], [a].[Id], [t].[ApplicationUserId], [t].[OrganizationId], [t1].[OrganizationId], [t1].[LocationId], [t1].[CreatedBy], [t1].[CreatedDateTime], [t1].[Default], [t1].[Deleted], [t1].[UpdatedBy], [t1].[UpdatedDateTime], [t1].[Id], [t1].[AddressLine1], [t1].[AddressLine2], [t1].[City], [t1].[CreatedBy0], [t1].[CreatedDateTime0], [t1].[Deleted0], [t1].[Fax], [t1].[Name], [t1].[Phone], [t1].[PostalCode], [t1].[PrivateCollection], [t1].[Province], [t1].[UpdatedBy0], [t1].[UpdatedDateTime0], [t2].[Id], [t2].[CreatedBy], [t2].[CreatedDateTime], [t2].[DateOfBirth], [t2].[Deleted], [t2].[FirstName], [t2].[Gender], [t2].[LastName], [t2].[OrgEmployeeId], [t2].[OrganizationId], [t2].[Suspended], [t2].[UpdatedBy], [t2].[UpdatedDateTime]
FROM [AspNetUsers] AS [a]
INNER JOIN (
    SELECT [a0].[ApplicationUserId], [a0].[OrganizationId]
    FROM [ApplicationUserOrganization] AS [a0]
    WHERE [a0].[Deleted] <> CAST(1 AS bit)
) AS [t] ON [a].[Id] = [t].[ApplicationUserId]
INNER JOIN (
    SELECT [o].[Id], [o].[Abbreviation], [o].[Active], [o].[CreatedBy], [o].[CreatedDateTime], [o].[Deleted], [o].[MaxTests], [o].[MinTests], [o].[Name], [o].[OrgGraphic], [o].[ProviderId], [o].[UpdatedBy], [o].[UpdatedDateTime]
    FROM [Organization] AS [o]
    WHERE [o].[Deleted] <> CAST(1 AS bit)
) AS [t0] ON [t].[OrganizationId] = [t0].[Id]
LEFT JOIN (
    SELECT [o0].[OrganizationId], [o0].[LocationId], [o0].[CreatedBy], [o0].[CreatedDateTime], [o0].[Default], [o0].[Deleted], [o0].[UpdatedBy], [o0].[UpdatedDateTime], [l].[Id], [l].[AddressLine1], [l].[AddressLine2], [l].[City], [l].[CreatedBy] AS [CreatedBy0], [l].[CreatedDateTime] AS [CreatedDateTime0], [l].[Deleted] AS [Deleted0], [l].[Fax], [l].[Name], [l].[Phone], [l].[PostalCode], [l].[PrivateCollection], [l].[Province], [l].[UpdatedBy] AS [UpdatedBy0], [l].[UpdatedDateTime] AS [UpdatedDateTime0]
    FROM [OrganizationLocation] AS [o0]
    INNER JOIN [Location] AS [l] ON [o0].[LocationId] = [l].[Id]
    WHERE [o0].[Deleted] <> CAST(1 AS bit)
) AS [t1] ON [t0].[Id] = [t1].[OrganizationId]
LEFT JOIN (
    SELECT [e].[Id], [e].[CreatedBy], [e].[CreatedDateTime], [e].[DateOfBirth], [e].[Deleted], [e].[FirstName], [e].[Gender], [e].[LastName], [e].[OrgEmployeeId], [e].[OrganizationId], [e].[Suspended], [e].[UpdatedBy], [e].[UpdatedDateTime]
    FROM [Employee] AS [e]
    WHERE ([e].[Deleted] <> CAST(1 AS bit)) AND ([e].[Deleted] <> CAST(1 AS bit))
) AS [t2] ON [t0].[Id] = [t2].[OrganizationId]
WHERE [a].[Id] = @__userId_0
ORDER BY [a].[Id], [t].[ApplicationUserId], [t].[OrganizationId], [t0].[Id], [t1].[OrganizationId], [t1].[LocationId], [t1].[Id], [t2].[LastName], [t2].[FirstName], [t2].[Id]`


Basically I executed the SAME software operation and the SAME LINQ query!

**Comment by ajcvickers** [comment]
@fabiano-ferreira Can you share the debug level logging output? (See https://docs.microsoft.com/en-us/ef/core/logging-events-diagnostics/ for information about enabling logs.)

**Comment by fabiano-ferreira** [comment]
> @fabiano-ferreira Can you share the debug level logging output? (See https://docs.microsoft.com/en-us/ef/core/logging-events-diagnostics/ for information about enabling logs.)

Hi @ajcvickers !

Since this is happening not in Development environment, but in UAT environment, could you, please, provide some examples about how to do this EF Core log to a text file?

**Comment by smitpatel** [comment]
@fabiano-ferreira - When you say that you ran same LINQ query and it generated 2 different SQL but at the same time, in your comments you are asserting that first query is without any browser interaction and 2nd one is when refreshing your browser. Can you please elaborate how are you running those queries in both instances. Specifically if they don't trigger when doing same user action then they are following different codepath which can contribute to the issue.

**Comment by fabiano-ferreira** [comment]
> @fabiano-ferreira - When you say that you ran same LINQ query and it generated 2 different SQL but at the same time, in your comments you are asserting that first query is without any browser interaction and 2nd one is when refreshing your browser. Can you please elaborate how are you running those queries in both instances. Specifically if they don't trigger when doing same user action then they are following different codepath which can contribute to the issue.

HI @smitpatel !

Well, the front end application is a Blazor Server app. I'm using Repository pattern with more 2 layers beside the Front End Layer.
There's absolutely zero chance of having 2 different code paths.

I executed the action after a click of a button (Blazor server app click event). After that I run a method to retrieve some records to my web page. Basically the button click and conditions are EXACTLY the same.

In the end this button click calls a method in my Bll component which executes the SAME LINQ code:

`var userOrganizations = await _applicationUserRepository.GetAll()
                                        .Include(o => o.ApplicationUserOrganizations)
                                            .ThenInclude(o => o.Organization)
                                                .ThenInclude(o => o.OrganizationLocations).ThenInclude(o => o.Location)
                                        .Include(o => o.ApplicationUserOrganizations)
                                            .ThenInclude(o => o.Organization)
                                                .ThenInclude(o => o.Employees.OrderBy(l => l.LastName).ThenBy(f => f.FirstName))
                                        .Where(u => u.Id == userId)
                                    .SelectMany(o => o.ApplicationUserOrganizations)
                                    .Select(o => o.Organization)
                                    .ToListAsync();``

The query filters are being configures on my Context class (OnModelCreating method):

` modelBuilder.Entity<Employee>().HasQueryFilter(p => !p.Deleted);
            modelBuilder.Entity<ReqOrdering>().HasQueryFilter(p => !p.Deleted);
            modelBuilder.Entity<ReqOrderingEmployee>().HasQueryFilter(p => !p.Deleted);
            modelBuilder.Entity<OrganizationLocation>().HasQueryFilter(p => !p.Deleted);
            modelBuilder.Entity<Provider>().HasQueryFilter(p => !p.Deleted);
            modelBuilder.Entity<ReqOrderingEmployeeDetail>().HasQueryFilter(p => !p.Deleted);`

My code was already stable. I NEVER had this issue before. It ONLY started to happen after I updated my Dal layer to user EF Core 5.0.2.

**Comment by ajcvickers** [comment]
@fabiano-ferreira There is an example in the docs showing simple logging to a file.

**Comment by fabiano-ferreira** [comment]
Hello guys!

I didn't have the opportunity to log EF Core steps. But I just rolled back EF Core to version 5.0.1 and the problem disappeared.

In case you want to simulate the problem you can do the following:

- Create a Blazor Server application with a database access being done by EF Core 5.0.2 using SQL Server.
- On the DbContext class apply some query filter to a table
- Create a simple Blazor page with a button
- On the button click do some access to the table that have the query filter set
- Publish the application on IIS
- Open SQL Server Profiler and do a trace into the database being accessed
- Open the application on IIS and click the button
- You will realize that the filter is applied correctly
- Let the page idle with no access and the browser open for 1 day
- Do the button click again and check the SQL Server profile for the query
- You will realize that the query filter is not applied anymore
- Again: this is ONLY happening with EF Coe 5.0.2

**Comment by AndriySvyryd** [comment]
@fabiano-ferreira Please attach the project you created following the steps above to make sure we are looking at the same code.

**Comment by fabiano-ferreira** [comment]
Hello guys!

You can close the ticket, please. After a lot of investigation I found that the error that was causing the Query filter to be ignored was inside my application. It was a very specific situation leading to this bug.

Thanks anyway for all your help!

Fabiano
